File: frontend/transcription-frontend/src/lib/api/notifications.ts
// Programmer Name : Mr. Reehan Mohamed Ashraf, TP077077, APD3F2511SE, Software Engineering Student, APU, Technology Park Malaysia
// Program Name: notifications.ts
// Description: API client for notifications
// First Written on: 03/07/2026
// Edited on: 21/07/2026
import { request } from "./client";
import type { NotificationListResponse, NotificationReadAllResponse, NotificationReadResponse, NotificationUnreadCountResponse } from "./types";

export function listNotifications(params: { page?: number; pageSize?: number; unreadOnly?: boolean } = {}, signal?: AbortSignal) {
  const query = new URLSearchParams();
  query.set("page", String(params.page ?? 1));
  query.set("pageSize", String(params.pageSize ?? 10));
  if (params.unreadOnly) query.set("unreadOnly", "true");
  return request<NotificationListResponse>(`/api/notifications?${query.toString()}`, { signal });
}

export function getNotificationUnreadCount(signal?: AbortSignal) {
  return request<NotificationUnreadCountResponse>("/api/notifications/unread-count", { signal });
}

export function markNotificationRead(notificationId: string, signal?: AbortSignal) {
  return request<NotificationReadResponse>(`/api/notifications/${encodeURIComponent(notificationId)}/read`, { method: "POST", signal });
}

export function markAllNotificationsRead(signal?: AbortSignal) {
  return request<NotificationReadAllResponse>("/api/notifications/read-all", { method: "POST", signal });
}

File: backend/internal/handlers/notifications.go
// Programmer Name : Mr. Reehan Mohamed Ashraf, TP077077, APD3F2511SE, Software Engineering Student, APU, Technology Park Malaysia
// Program Name: notifications.go
// Description: HTTP handler for notifications
// First Written on: 03/07/2026
// Edited on: 21/07/2026
package handlers

import (
	"net/http"
	"strconv"
	"strings"

	"transcript_app/backend/internal/dtos"
	"transcript_app/backend/internal/services"

	"github.com/gin-gonic/gin"
)

func APIListNotifications(c *gin.Context) {
	user, ok := CurrentUser(c)
	if !ok {
		writeUnauthenticated(c)
		return
	}
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("pageSize", "20"))
	unreadOnly := strings.EqualFold(c.Query("unreadOnly"), "true") || c.Query("unreadOnly") == "1"
	response, err := services.ListNotifications(c.Request.Context(), user.ID, services.NotificationListOptions{Page: page, PageSize: pageSize, UnreadOnly: unreadOnly})
	if err != nil {
		writeServiceError(c, err)
		return
	}
	c.JSON(http.StatusOK, response)
}

func APIGetUnreadNotificationCount(c *gin.Context) {
	user, ok := CurrentUser(c)
	if !ok {
		writeUnauthenticated(c)
		return
	}
	count, err := services.GetUnreadCount(c.Request.Context(), user.ID)
	if err != nil {
		writeServiceError(c, err)
		return
	}
	c.JSON(http.StatusOK, dtos.NotificationUnreadCountResponse{Count: count})
}

func APIMarkNotificationRead(c *gin.Context) {
	user, ok := CurrentUser(c)
	if !ok {
		writeUnauthenticated(c)
		return
	}
	notificationID := strings.TrimSpace(c.Param("notificationId"))
	if notificationID == "" {
		writeAPIError(c, http.StatusBadRequest, services.ErrCodeBadRequest, "notificationId is required", nil)
		return
	}
	notification, err := services.MarkRead(c.Request.Context(), user.ID, notificationID)
	if err != nil {
		writeServiceError(c, err)
		return
	}
	c.JSON(http.StatusOK, dtos.NotificationReadResponse{Notification: notification})
}

func APIMarkAllNotificationsRead(c *gin.Context) {
	user, ok := CurrentUser(c)
	if !ok {
		writeUnauthenticated(c)
		return
	}
	updated, err := services.MarkAllRead(c.Request.Context(), user.ID)
	if err != nil {
		writeServiceError(c, err)
		return
	}
	c.JSON(http.StatusOK, dtos.NotificationReadAllResponse{Updated: updated})
}
